app/models/concerns/variant_unique_sku_validation.rb

```diff
@@ -0,0 +1,29 @@
+# This adds validation to require SKU to be unique among the non-master and not deleted variants of
+# the same product. Additionally, there can be at most one variant with blank SKU for a products
+# variants with the same unit_value, unit_description, display_name, and display_as.
+#
+# Master and deleted variants are ignored.
+
+module VariantUniqueSkuValidation
+  extend ActiveSupport::Concern
+
+  included do
+    # Including is_master and deleted_at in the uniqueness scope is a workaround to ignore master
+    # and deleted variants.
+    #
+    # Handle validation for blank SKU separately.
+    validates :sku, uniqueness: { scope: [:product_id, :is_master, :deleted_at], allow_blank: true,
```

Comment:
So two different products can have variants with the same SKU? Do we want that? Or do we want all variants of one producer to have unique SKUs?

Reply:
Are there invalid variants in our production data already? Then we need a migration.